### ISSUE TITLE
ASpIRE data augmentation issue : Bug in wav-reverberate command

### ISSUE THREAD

**Issue by vijayaditya** [kaldi-asr/kaldi#889]
Using the latest data-augmentation scripts which use the `wav-reverberate` command I am unable to replicate the results in the ASpIRE recipe. When I use the old data-augmentation scripts from around May 2015, which use the older `local/multi_condition/corrupt.py` script, I seem to be getting better train/validation probabilities. 

This issue has been raised to keep track of the progress of this investigation. 

@tomkocse This might interest you too, as you reported similar issue few months back.

**Comment by vijayaditya** [comment]
I can confirm that the new data-augmentation recipe is not working. After 100/1000 iterations of training using TDNNs there is 25% rel. difference in WERs. I will now investigate what is the exact issue in the new augmentation recipe.

old data-augmentation 
```
%WER 34.2 | 2120 27212 | 71.6 18.3 10.2 5.8 34.2 80.2 | -0.613 | exp/nnet3/tdnn/decode_dev_aspire_whole_uniformsegmented_win10_over5_v6_200jobs_iter100_pp_fg/score_14/penalty_0.0/ctm.filt.filt.sys
```

new data-augmentation
```
%WER 45.4 | 2120 27217 | 60.2 24.7 15.1 5.6 45.4 85.5 | -0.645 | exp/nnet3/tdnn/decode_dev_aspire_whole_uniformsegmented_win10_over5_v6_200jobs_iter100_pp_fg/score_11/penalty_0.0/ctm.filt.filt.sys
```

**Comment by danpovey** [comment]
If the difference is that bad, I would have thought that the difference
would be clear from listening to it.
Are you sure it's that, though?  Could it be something went wrong in the
training?  Are the final objf values OK looking?
Dan

**Comment by vijayaditya** [comment]
The objfns look significantly better with the old augmentation recipe (see plot below and the attached pdf with more plots). I have used the same scripts for both these experiments. Only the feature directory `data/train_rvb_hires` is different.

I will perform a manual verification of the reverberated speech. 

![image](https://cloud.githubusercontent.com/assets/3895456/16715335/961904f0-46aa-11e6-94e1-5b2a87d55039.png)



[report.pdf](https://github.com/kaldi-asr/kaldi/files/356140/report.pdf)

**Comment by vijayaditya** [comment]
Confirmed bug in `wav-reverberate` command. It is not shifting the reverberated waveform to ensure that it is aligned with the input. This can be done by finding the position of the peak of the impulse response in the RIR file and shifting the reverberated waveform by this amount. This shifting is necessary to make the alignments from the clean speech waveforms useful. In some RIR files the peak can be as much as few hundred milliseconds away from the beginning so this is very critical stage.

In the below figure row 1 corresponds to the output from the `wav-reverberate` command; row 2 corresponds to the output from `aspire/s5/local/multi_condition/corrupt.py` script; and row 3 corresponds to the clean speech file. The shift in the output of `wav-reverberate` command is clearly visible.
![image](https://cloud.githubusercontent.com/assets/3895456/16715594/918b38d8-46b2-11e6-95bc-a0a4d37b4024.png)

As there could be severe misalignment errors the worse objfns we observe are understandable.

**Comment by vijayaditya** [comment]
@tomkocse Could you please add a new feature to the `wav-reverberate` command. If option `--shift-output` is true, find the peak of the impulse response and shift the output by the required amount.

**Comment by vijayaditya** [comment]
@xiaohui-zhang, @jtrmal  This might affect you too please verify if you are using `wav-reverberate` in your data augmentation recipe.

**Comment by danpovey** [comment]
BTW, I think it would make sense, after shifting, to allow the impulse
response you're convolving with to have nonzero values for negative t.
This would probably more closely approximate "real" impulse responses,
which won't always start with their maximum value at the most negative time.

**Comment by vijayaditya** [comment]
The recommendation is to shift the reverberated output rather than the impulse response as almost all the impulse responses I had seen had non-zero values for a few tens of milliseconds before the peak.

**Comment by tomkocse** [comment]
Ok

**Comment by tomkocse** [comment]
@vijayaditya In our discussion in #716 , the usage of wav-reverberate is going to be changed. Do you want me to to commit the whole new wav-reverberate or just fix the bug now? 

BTW, may i know whether you have explicitly done the shift in your older setup (local/multi_condition/corrupt.py)?

**Comment by vijayaditya** [comment]
@tomkocse As the new `wav-reverberate` binary is near completion you could combine this bug-fix with that PR. 

I did an explicit shift see this [line](https://github.com/kaldi-asr/kaldi/blob/master/egs/aspire/s5/local/multi_condition/corrupt.py#L116)

**Comment by vijayaditya** [comment]
Using the data augmentation scripts I was able to reproduce the nnet2 results, using new nnet3 models.

```
nnet2: 
%WER 30.8 | 2120 27213 | 75.3 16.2 8.4 6.2 30.8 78.8 | -0.724 | exp/nnet2_multicondition/nnet_ms_a/decode_dev_aspire_whole_uniformsegmented_win10_over5_v6_iterfinal_pp_fg/score_13/penalty_0.0/ctm.filt.filt.sys

nnet3:
%WER 31.0 | 2120 27217 | 74.8 16.1 9.1 5.9 31.0 77.9 | -0.707 | exp/nnet3/tdnn/decode_dev_aspire_whole_uniformsegmented_win10_over5_v6_200jobs_iterfinal_pp_fg/score_14/penalty_0.0/ctm.filt.filt.sys

```

**Comment by tomkocse** [comment]
@vijayaditya are you already using the wav-reverberate.cc in #706 ?

**Comment by vijayaditya** [comment]
@tomkocse no these are results using corrupt.py

**Comment by vijayaditya** [comment]
Corrected in #706 .
